buildSrc/src/main/groovy/airbyte-docker.gradle

```diff
@@ -0,0 +1,216 @@
+                if(!currentBase.equals(storedBase)) {
+                    logger.debug "did not match"
+                    return true
+                } else {
+                    logger.debug "did match"
+                    return false
+                }
+            }
+
+            if(notUpToDate) {
+                return false;
+            }
+
+            logger.debug "stored " + stored.trim()
+
+            def upToDate = current.equals(stored.trim())
+
+            logger.debug "uptodate " + upToDate.toString()
+
+            return upToDate
+        } else {
+            return false
+        }
+    }
+
+    static def createTask(Project project, String taskName, String dockerFile) {
+        if (project.file(dockerFile).exists()) {
+            def filteredProjectFiles = project.fileTree(project.projectDir).filter {
+                file -> !file.toString().contains(".venv")
+            }
+            project.task(taskName, type: AirbyteDockerTask) {
+                def dockerPath = Paths.get(project.projectDir.absolutePath, dockerFile)
+                def hash = MessageDigest.getInstance("MD5").digest(dockerPath.getBytes()).encodeHex().toString()
+                rootDir = project.rootProject.rootDir
+                projectDir = project.projectDir
+                projectFiles = filteredProjectFiles
+                idFileOutput = project.file(Paths.get(project.rootProject.rootDir.absolutePath, '.dockerversions', hash).toString())
+                baseImageHashes = getBaseImageHashes(project.rootProject.imageToVersion, dockerPath.toFile())
+                dependsOn project.assemble
+
+                outputs.upToDateWhen {
+                    return isUpToDate(logger, idFileOutput, project, dockerPath)
+                }
+            }
+        } else {
+            project.task(taskName) {
+                logger.info "Skipping ${taskName} because ${dockerFile} does not exist."
+            }
+        }
+    }
+
+    void apply(Project project) {
+        // set (and cache) global image to version map
+        project.rootProject.ext.imageToVersion = {
+            if(!project.rootProject.hasProperty("imageToVersion")) {
+                def imageToVersion = [:]
+                def stdout = new ByteArrayOutputStream()
+                project.exec {
+                    commandLine "docker", "images", "--no-trunc", "-f", "dangling=false", "--format", "{{.Repository}}:{{.Tag}} {{.ID}}"
+                    standardOutput = stdout;
+                }
+
+                "$stdout".eachLine { line ->
+                    def splits = line.split()
+                    imageToVersion.put(splits[0], splits[1])
+                }
+
+                return imageToVersion
+            } else {
+                return project.rootProject.imageToVersion
+            }
+        }()
+
+        createTask(project, 'airbyteDockerBuildMain', 'Dockerfile')
+        createTask(project, 'airbyteDockerBuildTest', 'Dockerfile.test')
```

Comment:
this task won't exist for every project which builds images. Could we instead invoke the plugin with a custom task name which takes as input the name of the dockerfile? e.g the calling project's build.gradle would have a block: 

```
task buildImage(type: AirbyteDocker){
  imageName = 'Dockerfile'
}

task buildTestImage(type: AirbyteDocker){
  imageName = 'Dockerfile.test'
}
```

It's also a lot clearer in the calling context that we are defining a buildImage task. Right now it's unclear how our custom files influence the calling gradle.build as it requires reading the imported file to understand what tasks it adds.

Reply:
The problem with doing it this way is that `buildImage` and `buildTestImage` then need dependencies configured at the project level, which I'd like to avoid. We could have separate plugins `airbyteDocker` and `airbyteDockerTest` that default to `Dockerfile` and `Dockerfile.test`. I don't know if that provides much value over what I have now though if we allow overriding the main and test dockerfile names. I'm not sure if we should do that unless it's actually needed for something, so we can encourage this convention.